clients/adapter-amqp/src/main/java/org/eclipse/hono/adapter/client/amqp/AbstractRequestResponseServiceClient.java

```diff
@@ -0,0 +1,369 @@
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.adapter.client.amqp;
+
+import java.net.HttpURLConnection;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+
+import org.apache.qpid.proton.amqp.messaging.ApplicationProperties;
+import org.apache.qpid.proton.message.Message;
+import org.eclipse.hono.cache.CacheProvider;
+import org.eclipse.hono.cache.ExpiringValueCache;
+import org.eclipse.hono.client.HonoConnection;
+import org.eclipse.hono.client.RequestResponseClientConfigProperties;
+import org.eclipse.hono.client.SendMessageSampler;
+import org.eclipse.hono.client.ServiceInvocationException;
+import org.eclipse.hono.client.StatusCodeMapper;
+import org.eclipse.hono.client.impl.CachingClientFactory;
+import org.eclipse.hono.config.ProtocolAdapterProperties;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.CacheDirective;
+import org.eclipse.hono.util.MessageHelper;
+import org.eclipse.hono.util.RequestResponseResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.opentracing.Span;
+import io.opentracing.tag.Tags;
+import io.vertx.core.Future;
+import io.vertx.core.buffer.Buffer;
+
+/**
+ * A vertx-proton based parent class for the implementation of API clients that follow the request response pattern.
+ * <p>
+ * Provides access to a {@link CacheProvider} which can be used to create caches for service response messages.
+ *
+ * @param <R> The type of response this client expects the peer to return.
+ * @param <T> The type of object contained in the peer's response.
+ *
+ */
+public abstract class AbstractRequestResponseServiceClient<T, R extends RequestResponseResult<T>> extends AbstractServiceClient {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractRequestResponseServiceClient.class);
+    private static final int[] CACHEABLE_STATUS_CODES = new int[] {
+                            HttpURLConnection.HTTP_OK,
+                            HttpURLConnection.HTTP_NOT_AUTHORITATIVE,
+                            HttpURLConnection.HTTP_PARTIAL,
+                            HttpURLConnection.HTTP_MULT_CHOICE,
+                            HttpURLConnection.HTTP_MOVED_PERM,
+                            HttpURLConnection.HTTP_GONE
+    };
+
+    /**
+     * Contains the AMQP link sender/receiver link pairs for invoking the service.
+     */
+    protected final CachingClientFactory<RequestResponseClient<R>> clientFactory;
+    /**
+     * A cache to use for responses received from the service.
+     */
+    private final ExpiringValueCache<Object, R> responseCache;
```

Comment:
Not directly related to the changes here, but:
Have you considered letting the same cache instance be used for all client instances, i.e. across verticle instances?
Then we would not only get cached tenant values when a connection is down, but also when the tenant object has already been retrieved via another connection/verticle.
I guess from a thread-safety point of view, this would be ok using CaffeineCache.

Reply:
In general, that should already be possible because the verticle instances get access to the `ExpiringvalueCache` via a `CacheProvider`. If they are all configured to use the same `CacheProvider` instance, then they should all use the same cache instance as well.